opm/grid/cpgrid/CpGrid.cpp

```diff
@@ -1516,610 +2031,1764 @@ void CpGrid::addLgrsUpdateLeafView(const std::vector<std::array<int,3>>& cells_p
-            // above levelZeroToLGRsBoundaryCorners_oneToOne declaration. See "Attention" comment.
-            if (!isEquivCornerLevelZero_laysOnLgrBoundary || (lgrsToLevelZeroBoundaryCorners_oneToOne.count({patch+1, corner}) == 1) ){
-                level_to_leaf_corners[patch + 1][corner] = corner_count;
-                corner_count +=1;
+            for (const auto& [markedFace, itsRefinedFaces] : parentFace_to_itsRefinedFaces) {
+                faceInMarkedElemAndRefinedFaces[markedFace].push_back({elemIdx, itsRefinedFaces});
             }
+        } // end-if-elemMark==1
+    } // end-elem-for-loop
+}
+
+void CpGrid::defineChildToParentAndIdxInParentCell(std::vector<std::vector<std::array<int,2>>>& refined_child_to_parent_cells_vec,
+                                                   std::vector<std::vector<int>>& refined_cell_to_idxInParentCell_vec,
+                                                   std::vector<std::array<int,2>>& adapted_child_to_parent_cells,
+                                                   std::vector<int>& adapted_cell_to_idxInParentCell,
+                                                   std::map<std::array<int,2>,std::array<int,2>> refinedLevelAndRefinedCell_to_elemLgrAndElemLgrCell,
+                                                   const std::vector<int>& refined_cell_count_vec,
+                                                   std::unordered_map<int,std::array<int,2>> adaptedCell_to_elemLgrAndElemLgrCell,
+                                                   const int cell_count)
+{
+    const int& startingGridIdx = static_cast<int>(this->data_.size()) -1;
+    adapted_child_to_parent_cells.resize(cell_count, std::array<int,2>{-1,-1}); //  Entry is {-1,-1} when cell has no father, otherwise, {level parent cell, parent cell index}.
+    adapted_cell_to_idxInParentCell.resize(cell_count, -1);
+    // Rewrite only the entries of adapted cells that have a parent cell
+    for (int cell = 0; cell < cell_count; ++cell) {
+        const auto& elemLgr = adaptedCell_to_elemLgrAndElemLgrCell[cell][0];
+        // (elemLgr == -1) represents that the adapted cell is equivalent to a cell from the starting grid (current_view_data_)
+        if (elemLgr != -1) {
+            // Search for the level where the parent cell was born, and its index in that level grid.
+            // Notice that elemLgr is a cell index in current_view_data_ (the starting grid where elements got marked for (further) refinement).
+            const auto& element = Dune::cpgrid::Entity<0>(*current_view_data_, elemLgr, true); // elemLgr == parent cell index in starting grid.
+            const auto& elemLevel = element.level(); // parent cell level (where the parent cell was born, might not coincide with the starting grid).
+            const auto& elemLevelIdx = (elemLevel == 0) ? element.getOrigin().index() : element.getLevelElem().index(); // parent cell index in the level it was born.
```

Comment:
I see your point. Unfortunately, getOrigin() returns 
- the parent entity if hasFather(), or,
- for cells on the leaf-grid-view, without father, returns the equivalent entity on the level that entity was born [To be fixed now: current code only considers level 0 as possible grids. I'll change this!], otherwise, 
- returns itself. g
getLevelElem() instead, it's supposed to be used only for leaf-grid-view cells, and it returns the equivalent refined or coarse cell on the level-grid where it was born. 

So, I would say you're right and we can simplify that piece of code, however, getOrigin() needs to consider all possible levels. I'll fix that. Great suggestion!

Reply:
I did simplify a bit this piece of code, however, I couldn't get rid of some if-statements. The main reason behind this is that we do not want always a father of an equivalent level element. For example, all the coarse cells not involved in any refinement (neither pre-adapt refined-grids nor new-refined-grids created in the current call of adapt) will get a child_to_parent_ value. They are "by default" {-1,-1}.